account_index_based_currency/models/account_invoice.py

```diff
@@ -0,0 +1,144 @@
+                self.company_id, self.date_invoice or fields.Date.today())
+        self.index_based_currency_amount = index_based_currency_amount
+        self.agreement_currency_amount = (
+            self.index_based_currency_amount / self.agreement_currency_rate)
+
+    @api.multi
+    @api.onchange('agreement_currency_rate')
+    def onchange_agreement_currency_rate(self):
+        if self.agreement_currency_rate <= 0:
+            raise UserError(_('Invalid currency rate value'))
+
+    @api.multi
+    @api.onchange('agreement_currency_id')
+    def onchange_agreement_currency_id(self):
+        self.agreement_currency_rate = self.agreement_currency_id._convert(
+            1, self.index_based_currency_id,
+            self.company_id,
+            self.date_invoice or fields.Date.today(), round=False)
+        self.agreement_currency_amount = self.index_based_currency_id._convert(
+            self.index_based_currency_amount, self.agreement_currency_id,
+            self.company_id,
+            self.date_invoice or fields.Date.today(), round=False)
+
+    @api.model
+    def default_get(self, default_fields):
+        res = super(AccountInvoice, self).default_get(default_fields)
+        default_rate = (
+            self.env.user.company_id.currency_id._convert(
+                1, self.env.user.company_id.index_based_currency_id,
+                self.env.user.company_id,
+                self.date_invoice or fields.Date.today(), round=False))
+        res['agreement_currency_id'] = self.env.user.company_id.currency_id.id
+        res['agreement_currency_rate'] = default_rate
+        return res
+
+    # old_currency_id = fields.Many2one(
+    #     'res.currency', help="The previous currency used")
+    agreement_currency_id = fields.Many2one(
+        'res.currency',
+        copy=True,
+        help="Currency at which the agreement is to be/was settled",
+        readonly=True, states={'draft': [('readonly', False)]},
+    )
+    index_based_currency_id = fields.Many2one(
+        'res.currency',
+        related='company_id.index_based_currency_id',
+        help="Currency used for reporting purposes",
+        copy=False,
+        readonly=True)
+    index_based_currency_amount = fields.Monetary(
+        currency_field='index_based_currency_id',
+        store=True, readonly=True, compute='_compute_currency_amount',
+        copy=False,
+        help="Total amount in the currency of the company, negative for "
+        "credit notes.")
+    agreement_currency_amount = fields.Monetary(
+        currency_field='agreement_currency_id',
+        store=True, readonly=True, compute='_compute_currency_amount',
+        copy=False,
+        help="Total amount in the currency of the company, negative for "
+        "credit notes.")
+    index_based_currency_rate = fields.Float(
+        default=1,
+        store=True, readonly=True, compute='_compute_currency_rate',
```

Comment:
@hbto The computed fields are readonly by default.
```suggestion
        store=True, compute='_compute_currency_rate',
```